shared/lib/neighbors/metrics.test.ts

```diff
@@ -0,0 +1,198 @@
+/**
+*  @license
+* Copyright 2021, JsData. All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { assert } from 'chai'
+import * as fc from 'fast-check'
+import { describe, it } from 'mocha'
+import { Metric, minkowskiDistance } from './metrics'
+
+const NDIMS = Object.freeze([1, 2, 7])
+
+const anyVec = (ndim: number) => {
+  // small value to provoke underflow (nextDown(x)**2 === 0)
+  const small = 1.5717277847026288e-162
+
+  // large value to provoke overflow (x*x === Infinity)
+  const large = 1.3407807929942597e154
+
+  const scale = 2 ** 16
+
+  const anyDouble = fc.oneof(
+    fc.double(-scale, +scale),
+    fc.oneof(
+      fc.double(0, +small * scale), // <- underflow
+      fc.double(-small * scale, 0), // <- underflow
+      fc.double(large / scale, +Number.MAX_VALUE), // <- overflow
+      fc.double(-Number.MAX_VALUE, large / scale) // <- overflow
+    )
+  )
+  return fc.array(anyDouble, { minLength: ndim, maxLength: ndim })
+}
```

Comment:
I think these fast-check tests are really cool @DirkToewe. Seriously I think this is gonna be huge :D

Reply:
The lazy man's way to create test cases :laughing: